harness/go/history/history.go

```diff
@@ -0,0 +1,252 @@
+}
+
+// UnmarshalJSON converts the given JSON to histories.
+func (h *Histories) UnmarshalJSON(b []byte) error {
+	// Unmarshal into raw JSON array, then into each with proto unmarshaler
+	var halfUnmarshaled []json.RawMessage
+	if err := json.Unmarshal(b, &halfUnmarshaled); err != nil {
+		return err
+	}
+	hists := make([]*history.History, len(halfUnmarshaled))
+	for i, histJSON := range halfUnmarshaled {
+		var hist history.History
+		if err := jsonpb.Unmarshal(bytes.NewReader(histJSON), &hist); err != nil {
+			return err
+		}
+		hists[i] = &hist
+	}
+	*h = hists
+	return nil
+}
+
+var historyMarshaler jsonpb.Marshaler
+
+// MarshalJSON converts the histories to JSON.
+func (h Histories) MarshalJSON() ([]byte, error) {
+	// Copy and sort each history by its first event's name
+	sorted := make(Histories, len(h))
+	copy(sorted, h)
+	var err error
+	sort.Slice(sorted, func(i, j int) bool {
+		a, aErr := historyFirstEventName(sorted[i])
+		b, bErr := historyFirstEventName(sorted[j])
+		if err == nil && aErr != nil {
+			err = aErr
+		} else if err == nil && bErr != nil {
+			err = bErr
+		}
+		return a < b
+	})
+
+	// Marshal each history, then marshal the whole thing
+	halfMarshaled := make([]json.RawMessage, len(sorted))
+	for i, history := range sorted {
+		s, err := historyMarshaler.MarshalToString(history)
+		if err != nil {
+			return nil, fmt.Errorf("failed marshaling history: %w", err)
+		}
+		halfMarshaled[i] = json.RawMessage(s)
+	}
+	return json.Marshal(halfMarshaled)
+}
+
+// Clone performs a deep clone of histories.
+func (h Histories) Clone() Histories {
+	ret := make(Histories, len(h))
+	for i, hist := range h {
+		ret[i] = proto.Clone(hist).(*history.History)
+	}
+	return ret
+}
+
+// Equals checks history equality.
+func (h Histories) Equals(other Histories) bool {
+	if len(h) != len(other) {
+		return false
+	}
+	for i, hist := range h {
+		if !proto.Equal(hist, other[i]) {
+			return false
+		}
+	}
+	return true
+}
+
+// ScrubRunSpecificFields removes all fields on the history that are specific to
+// the run.
+func (h Histories) ScrubRunSpecificFields() {
+	scrubRunSpecificFields(reflect.ValueOf(h))
+}
+
+func scrubRunSpecificFields(v reflect.Value) {
+	if !v.IsValid() || v.IsZero() {
+		return
+	}
+	// First scrub the fields
+	scrubRunSpecificScalars(v.Interface())
+	// Now walk children and scrub
+	switch v.Kind() {
+	case reflect.Array, reflect.Slice:
+		for i := 0; i < v.Len(); i++ {
+			scrubRunSpecificFields(v.Index(i))
+		}
+	case reflect.Interface, reflect.Ptr:
+		scrubRunSpecificFields(v.Elem())
+	case reflect.Map:
+		iter := v.MapRange()
+		for iter.Next() {
+			scrubRunSpecificFields(iter.Value())
+		}
+	case reflect.Struct:
+		for i := 0; i < v.Type().NumField(); i++ {
+			if v.Type().Field(i).IsExported() {
+				scrubRunSpecificFields(v.Field(i))
+			}
+		}
+	}
+}
+
+func scrubRunSpecificScalars(v interface{}) {
+	// TODO: Add to this as more things are discovered
+	switch v := v.(type) {
```

Comment:
I'm not sure about this, but it honestly might be easier to say what *is* valid to compare.

Namely, even type and fields like ids, activity type. That's mostly it honestly

Reply:
This looks longer/grosser than it is to maintain. Once you start going a reflective route to either include or exclude specific ones, you quickly start getting more complex code to just walk the things and zero them out than to just do this.